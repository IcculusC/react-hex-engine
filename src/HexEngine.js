import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import { HexEngineProvider } from "./Context";
import Point from "./models/Point";
import Orientation from "./models/Orientation";

function getPointOffset(corner, orientation, size) {
  let angle = (2.0 * Math.PI * (corner + orientation.startAngle)) / 6;
  return new Point(size.x * Math.cos(angle), size.y * Math.sin(angle));
}

function calculateCoordinates(orientation, size) {
  const center = new Point(0, 0);
  return [...Array(6).keys()].map((_, i) => {
    const offset = getPointOffset(i, orientation, size);
    return new Point(center.x + offset.x, center.y + offset.y);
  });
}

const HexEngine = props => {
  const {
    children,
    classes,
    flat,
    height,
    origin,
    size,
    spacing,
    viewBox,
    width
  } = props;

  const orientation = flat ? Orientation.Flat : Orientation.Pointy;
  const [points, setPoints] = useState(
    calculateCoordinates(orientation, size)
      .map(point => point.toString())
      .join(" ")
  );
  useEffect(
    () => {
      setPoints(() =>
        calculateCoordinates(orientation, size)
          .map(point => point.toString())
          .join(" ")
      );
    },
    [orientation, size]
  );
  const layout = { orientation, origin, size, spacing };

  return (
    <svg
      key={JSON.stringify(layout)}
      className={classNames("grid", classes.grid)}
      height={height}
      version="1.1"
      viewBox={`${viewBox.x} ${viewBox.y} ${viewBox.width} ${viewBox.height}`}
      width={width}
      xmlns="http://www.w3.org/2000/svg"
    >
      <HexEngineProvider
        value={{
          layout,
          points,
          viewBox
        }}
      >
        <g className={classes.layout}>{children}</g>
      </HexEngineProvider>
    </svg>
  );
};

HexEngine.propTypes = {
  children: PropTypes.node.isRequired,
  classes: PropTypes.objectOf(PropTypes.string),
  /** Determines if the hexagons are oriented with a point or an edge facing up */
  flat: PropTypes.bool,
  /** CSS string or number */
  height: PropTypes.oneOfType([
    PropTypes.string.isRequired,
    PropTypes.number.isRequired
  ]),
  /** Origin of the grid */
  origin: PropTypes.oneOfType([
    PropTypes.instanceOf(Point),
    PropTypes.shape({
      x: PropTypes.number.isRequired,
      y: PropTypes.number.isRequired
    })
  ]),
  /** Size of the hexagons in each dimension */
  size: PropTypes.oneOfType([
    PropTypes.instanceOf(Point),
    PropTypes.shape({
      x: PropTypes.number.isRequired,
      y: PropTypes.number.isRequired
    })
  ]),
  /** Space between hexagons */
  spacing: PropTypes.number,
  /** CSS string or number */
  width: PropTypes.oneOfType([
    PropTypes.string.isRequired,
    PropTypes.number.isRequired
  ]),
  /** The viewBox {x,y,width,height} of the svg view area */
  viewBox: PropTypes.shape({
    height: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired,
    x: PropTypes.number.isRequired,
    y: PropTypes.number.isRequired
  })
};

HexEngine.defaultProps = {
  classes: { grid: "", layout: "" },
  flat: true,
  height: 480,
  origin: new Point(0, 0),
  size: new Point(10, 10),
  spacing: 1.0,
  width: 640,
  viewBox: {
    height: 100,
    width: 100,
    x: -50,
    y: -50
  }
};

export const HexEngine_ = HexEngine;
export default HexEngine_;
