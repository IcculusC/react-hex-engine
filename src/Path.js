import React, { useContext } from "react";
import PropTypes from "prop-types";
import HexUtils from "./HexUtils";
import { HexEngineContext } from "./Context";

function getPoints(start, end, layout) {
  if (!end || !start) {
    return "";
  }

  const distance = HexUtils.distance(start, end);
  const step = 1.0 / Math.max(distance, 1);
  const points = [...Array(distance + 1).keys()]
    .map(i => {
      const hex = HexUtils.round(HexUtils.hexLerp(start, end, step * i));
      const pixel = HexUtils.hexToPixel(hex, layout);
      return ` ${pixel.x},${pixel.y} `;
    })
    .join("L");

  return `M${points}`;
}

export const Path = ({ start, end }) => {
  const { layout } = useContext(HexEngineContext);
  return <path d={getPoints(start, end, layout)} />;
};

Path.propTypes = {
  end: PropTypes.object,
  layout: PropTypes.object,
  start: PropTypes.object
};

export const Path_ = Path;
export default Path_;
