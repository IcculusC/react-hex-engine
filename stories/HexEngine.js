import React from "react";
import { storiesOf } from "@storybook/react";
import { withKnobs, boolean, number, object } from "@storybook/addon-knobs";
import { withInfo } from "@storybook/addon-info";
import GridGenerator from "../src/GridGenerator";
import Hexagon, { Hexagon_ } from "../src/Hexagon";
import HexEngine from "../src/HexEngine";
import Point from "../src/models/Point";
import IndexedEngine from "../src/IndexedEngine";
import "./Hexagon.css";

Hexagon.displayName = "Hexagon";
Hexagon_.displayName = "Hexagon";
HexEngine.displayName = "HexEngine";

const stories = storiesOf("HexEngine", module);

stories.addDecorator(withKnobs);

stories.addDecorator(
  withInfo({
    header: false,
    inline: true,
    maxPropStringLength: 512,
    source: true,
    propTablesExclude: [Hexagon_, Hexagon]
  })
);

stories
  .add(
    "Kitchen Sink",
    () => {
      const flat = boolean("flat", true, "HexEngine");
      const size = object("size", { x: 10, y: 10 }, "HexEngine");
      const origin = object("origin", { x: 0, y: 0 }, "HexEngine");
      const spacing = number("spacing", 1.0, {}, "HexEngine");
      const width = number("width", 320, {}, "HexEngine");
      const height = number("height", 240, {}, "HexEngine");
      const viewBox = object(
        "viewBox",
        { x: -30, y: -30, width: 60, height: 60 },
        "HexEngine"
      );

      return (
        <HexEngine
          width={width}
          height={height}
          viewBox={viewBox}
          flat={flat}
          size={size}
          origin={origin}
          spacing={spacing}
        >
          {GridGenerator.hexagon(1).map(hex => (
            <Hexagon {...hex} key={hex.toString()} />
          ))}
        </HexEngine>
      );
    },
    {
      info: {
        propTables: [HexEngine]
      }
    }
  )
  .add("non zero origin", () => (
    <HexEngine
      origin={new Point(-5, -5)}
      width={320}
      height={240}
      viewBox={{ x: -30, y: -30, width: 60, height: 60 }}
    >
      <Hexagon q={0} r={0} s={0} />
    </HexEngine>
  ))
  .add("pointy orientation", () => (
    <HexEngine
      flat={false}
      width={320}
      height={240}
      viewBox={{
        x: -25,
        y: -25,
        width: 50,
        height: 50
      }}
    >
      <Hexagon q={0} r={0} s={0} />
    </HexEngine>
  ))
  .add("bigger size", () => (
    <HexEngine
      size={new Point(30, 30)}
      width={320}
      height={240}
      viewBox={{ x: -30, y: -30, width: 60, height: 60 }}
    >
      <Hexagon q={0} r={0} s={0} />
    </HexEngine>
  ))
  .add("smaller size", () => (
    <HexEngine
      size={new Point(3, 3)}
      width={320}
      height={240}
      viewBox={{ x: -30, y: -30, width: 60, height: 60 }}
    >
      <Hexagon q={0} r={0} s={0} />
    </HexEngine>
  ))
  .add("irregular proportions", () => (
    <HexEngine
      size={new Point(15, 5)}
      width={320}
      height={240}
      viewBox={{ x: -30, y: -30, width: 60, height: 60 }}
    >
      <Hexagon q={0} r={0} s={0} />
    </HexEngine>
  ))
  .add("spacing", () => (
    <HexEngine
      spacing={1.05}
      width={320}
      height={240}
      viewBox={{ x: -30, y: -30, width: 60, height: 60 }}
    >
      <Hexagon q={-1} r={0} s={1} />
      <Hexagon q={0} r={0} s={0} />
      <Hexagon q={1} r={0} s={-1} />
    </HexEngine>
  ))
  .add("indexed", () => {
    const viewBox = object(
      "viewBox",
      { x: -50, y: -50, width: 100, height: 100 },
      "HexEngine"
    );
    const map = GridGenerator.hexagon(50);
    const classes = {
      hexagon: "showCoordinates text",
      highlighted: "highlighted",
      hovered: "hovered",
      selected: "selected",
      q: "axis",
      r: "axis",
      s: "axis",
      text: "text"
    };
    return (
      <IndexedEngine
        height={400}
        map={map}
        viewBox={viewBox}
        width={400}
        spacing={1.05}
      >
        {hexes =>
          hexes.map(hex => (
            <Hexagon
              {...hex}
              classes={classes}
              showCoordinates
              selected={Math.random() > 0.5}
            />
          ))
        }
      </IndexedEngine>
    );
  });
