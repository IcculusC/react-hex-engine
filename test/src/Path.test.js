import React from "react";
import renderer from "react-test-renderer";

import HexEngine from "../../src/HexEngine";
import Path from "../../src/Path";

test("Path should render correctly", () => {
  const tree = renderer
    .create(
      <HexEngine
        classes={{ layout: "test1" }}
        size={{ x: 6, y: 6 }}
        flat={false}
        spacing={1.1}
        origin={{ x: 0, y: 0 }}
      >
        <Path start={{ q: 1, r: 1, s: -1 }} end={{ q: 0, r: 0, s: 0 }} />
      </HexEngine>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Path should render correctly without an end hex", () => {
  const tree = renderer
    .create(
      <HexEngine
        classes={{ layout: "test2" }}
        size={{ x: 6, y: 6 }}
        flat={false}
        spacing={1.1}
        origin={{ x: 0, y: 0 }}
      >
        <Path start={{ q: 1, r: 1, s: -1 }} />
      </HexEngine>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
