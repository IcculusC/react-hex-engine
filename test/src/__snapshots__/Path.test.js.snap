// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Path should render correctly 1`] = `
<g
  className="test1"
>
  <path
    d="M 17.147302994931888,9.9 L 0,0 "
  />
</g>
`;

exports[`Path should render correctly without an end hex 1`] = `
<g
  className="test2"
>
  <path
    d=""
  />
</g>
`;
